##!/bin/bash
## sorting of alignment files and log files generated by get_specific_alignments_pe.sh into subfolders
## Author: Doris Chen 
## version 240206

function usage()
{  echo
   echo "Sort specific alignment files"
   echo
   echo "Usage: $0 -1 GENOME1(cal.; ZP591.22|CBS138) -2 GENOME2(exp.; ASMv1|R64) -o OUTFILEBASE -c TRUE|FALSE"
	  echo
			echo "  -1 ... calibration genome; specific alignments are moved to 'SPECIES1_specific' subfolders"
			echo "  -2 ... experimental genome; specific alignments are moved to 'SPECIES2_specific' subfolders"
			echo "  -o ... counts are saved in 'OUTFILEBASE_specAlign-GENOME1-GENOME2_counts.txt'"
			echo "  -c ... if TRUE, only counts will be extracted, no sorting of files"
		    echo "  -h ... help"
			echo
			exit
}


## OPTIONS
while getopts ":1:2:o:c:h" OPTION; do
	 case $OPTION in
		  1 ) GENOME1="$OPTARG" ;;
		  2 ) GENOME2="$OPTARG" ;;
		  o ) OUTFILEBASE="$OPTARG" ;;
		  c ) COUNTONLY="$OPTARG" ;;
		  h ) usage ;;
		  * ) echo "Unrecognized argument. Use '-h' for usage information."
		  exit 1 ;;
	 esac
done

case "$GENOME1" in
		CBS138 ) SPECIES1=ngla ;;
	  ZP591.22 ) SPECIES1=skud ;;
			 * ) echo "Unknown genome "$GENOME1" !!"
			     exit 1 ;;
esac

case "$GENOME2" in
		   R64 ) SPECIES2=scer ;;
		 ASMv1 ) SPECIES2=scer ;;	 	  
			 * ) echo "Unknown genome "$GENOME2" !!"
			     exit 1 ;;
esac

OUTFILE=${OUTFILEBASE}_specAlign-${GENOME1}-${GENOME2}_counts.txt

## SCRIPT
# count specific experimental genome
echo "$GENOME2 specific:" | tee $OUTFILE
grep "_${GENOME1}_unaligned_${GENOME2}.sam aligned" *.log | cut -d':' -f2-3 | tee -a $OUTFILE  

# count specific calibration genome
echo "$GENOME1 specific:"  | tee -a $OUTFILE
grep "_${GENOME1}_woCommon-${GENOME2}.sam aligned" *.log | cut -d':' -f2-3  | tee -a $OUTFILE

# count common
echo "Common:"  | tee -a $OUTFILE
grep "_${GENOME1}_${GENOME2}.sam aligned" *.log | cut -d':' -f2-3  | tee -a $OUTFILE

# count unaligned
echo "Unknown:"  | tee -a $OUTFILE
grep "_${GENOME1}_unaligned_${GENOME2}_unaligned.sam unaligned" *.log | cut -d':' -f2-3  | tee -a $OUTFILE

# count raw number of aligned reads (from ngm)
echo "Raw aligned read counts (including improper pairs etc.):" | tee -a $OUTFILE
grep "reads mapped" *.log | cut -d' ' -f3-10 | sed 's/(//g' | sed 's/)//g' | tee -a $OUTFILE


if [ "$COUNTONLY" = "FALSE" ]; then
    mkdir -p ${SPECIES1}_specific
	mkdir -p ${SPECIES2}_specific
	mkdir -p common
	mkdir -p unknown
	mkdir -p log_files
	mkdir -p backup

	mv *_${GENOME1}_woCommon-${GENOME2}.sam ${SPECIES1}_specific
	mv *_${GENOME1}_unaligned_${GENOME2}.sam ${SPECIES2}_specific
	mv *_${GENOME1}_${GENOME2}.sam common
	mv *_${GENOME1}_unaligned_${GENOME2}_unaligned.sam unknown
	mv specAlign*.log log_files
	mv *_${GENOME1}.sam backup
fi

exit 0
